### PR TITLE
[#99] Add support for changing passwords (main)

### PR DESCRIPTION
**Pull request by korydraughn**

### CODE REVIEW

File: irods_specific_implementation/irods_rest_admin_api_implementation.h [hunk 1]
```diff
@@ -73,6 +93,38 @@ namespace irods::rest
...
+                // what is defined on the server-side.
+                std::strncat(plain_text_password.data(), "1gCBizHWbwIYyWLoysGzTe6SyzqFKMniZX05faZHWAwQKXf6Fs", lcopy);
+            }
+
+            std::array<char, MAX_PASSWORD_LEN + 10> admin_password{};
+
+            // Get the plain text password of the iRODS service account user.
+            // "obfGetPw" decodes the obfuscated password stored in .irods/.irodsA.
+            if (obfGetPw(admin_password.data()) != 0) {
+                THROW(SYS_INTERNAL_ERR, "failed to unobfuscate admin password in .irodsA file.");
```
Comment (korydraughn) [inline irods_specific_implementation/irods_rest_admin_api_implementation.h:113]: consider using a more descriptive error code

File: irods_specific_implementation/irods_rest_admin_api_implementation.h [hunk 2]
```diff
@@ -73,6 +93,38 @@ namespace irods::rest
...
+
+            debug("Client session signature = [{}]", getSessionSignatureClientside());
+
+            std::array<char, MAX_PASSWORD_LEN + 100> obfuscate_password{};
+            obfEncodeByKeyV2(plain_text_password.data(),
+                             admin_password.data(),
+                             getSessionSignatureClientside(),
+                             obfuscate_password.data());
+
+            debug("Obfuscated password = [{}]", std::string(obfuscate_password.data()));
```
Comment (korydraughn) [inline irods_specific_implementation/irods_rest_admin_api_implementation.h:124]: review need for `std::string()`

File: irods_specific_implementation/irods_rest_api_base.h [hunk 3]
```diff
@@ -232,7 +232,7 @@ namespace irods::rest
 
         std::string decode_url(const std::string& _in) const
         {
-            trace("Decoding input [{}] ...", _in);
+            trace("Decoding input [{}] ...", _in); // TODO This prints incoming arguments! (e.g. Passwords!)
```
Comment (korydraughn) [inline irods_specific_implementation/irods_rest_api_base.h:235]: perhaps this is why we left it as a trace message.

still, logging people's passwords is wrong. we should remove this line.

**Review by korydraughn — COMMENTED** [review]
don't forget to update the README about HTTPS and plaintext passwords.

File: irods_specific_implementation/irods_rest_admin_api_implementation.h (hunk 1 above)
Comment (trel) [reply irods_specific_implementation/irods_rest_admin_api_implementation.h:113]: nothing in the 900000 section seems quite correct.

nothing with 'auth' or 'pass' seems quite right.  

could be fine leaving this here.   nothing the user can do to fix this...

File: irods_specific_implementation/irods_rest_api_base.h (hunk 3 above)
Comment (trel) [reply irods_specific_implementation/irods_rest_api_base.h:235]: left note on other PR... just comment this out, along with a note as to why - available for devs to turn it back on if helpful.   but can never write in production.

File: irods_specific_implementation/irods_rest_admin_api_implementation.h (hunk 1 above)
Comment (korydraughn) [reply irods_specific_implementation/irods_rest_admin_api_implementation.h:113]: sure. i didn't spot anything in the error code table that seemed much better.

leaving as is.

File: irods_specific_implementation/irods_rest_api_base.h (hunk 3 above)
Comment (korydraughn) [reply irods_specific_implementation/irods_rest_api_base.h]: left a comment about why it is disabled.

File: irods_specific_implementation/irods_rest_admin_api_implementation.h (hunk 2 above)
Comment (korydraughn) [reply irods_specific_implementation/irods_rest_admin_api_implementation.h]: removed.

**Comment by korydraughn** [comment]
added test.

**Comment by korydraughn** [comment]
added #

**Pull request merged**